File: BEmu/IntradayTickRequest/ElementIntradayTickInt.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu" file="IntradayTickRequest/ElementIntradayTickInt.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace BEmu.IntradayTickRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class ElementIntradayTickInt : Element
    {
        private readonly int _value;
        private readonly string _name;

        internal ElementIntradayTickInt(string name, int value)
        {
            this._name = name;
            this._value = value;
        }

        public override Name Name { get { return new Name(this._name); } }
        public override int NumValues { get { return 1; } }
        public override int NumElements { get { return 0; } }
        public override bool IsComplexType { get { return false; } }
        public override bool IsArray { get { return false; } }
        public override bool IsNull { get { return false; } }
        public override IEnumerable<Element> Elements { get { if (this.NumValues < -1) yield return null; } } //This is the only way I can think to have the compiler output "Enumeration yielded no results"
        public override bool HasElement(string name, bool excludeNullElements = false) { return false; }

        public override object this[int index]
        {
            get
            {
                if (index == 0)
                    return this._value;
                else
                    return base[index];
            }
        }

        internal override StringBuilder PrettyPrint(int tabIndent)
        {
            return base.PrettyPrintHelper(tabIndent, this._value.ToString());
        }

        public override object GetValue()
        {
            return this._value;
        }

        public override object GetValue(int index)
        {
            if (index == 0)
                return this.GetValue();
            else
                return base.GetValue(index);
        }

        public override int GetValueAsInt32()
        {
            return this._value;
        }

        public override long GetValueAsInt64()
        {
            return (long)this._value;
        }

        public override float GetValueAsFloat32()
        {
            return (float)this._value;
        }

        public override double GetValueAsFloat64()
        {
            return (double)this._value;
        }

        public override string GetValueAsString()
        {
            return this._value.ToString();
        }
    }
}

File: BEmu_cpp/src/IntradayTickRequest/ElementIntradayTickInt.cpp
//------------------------------------------------------------------------------------------------
// <copyright project="BEmu_cpp" file="src/IntradayTickRequest/ElementIntradayTickInt.cpp" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------------------------

#include "IntradayTickRequest/ElementIntradayTickInt.h"
#include "BloombergTypes/Name.h"
#include "BloombergTypes/ElementPtr.h"

namespace BEmu
{
	namespace IntradayTickRequest
	{
		ElementIntradayTickInt::ElementIntradayTickInt(const std::string& name, int value)
		{
			this->_name = name;
			this->_value = value;
		}

		ElementIntradayTickInt::~ElementIntradayTickInt()
		{
		}

		Name ElementIntradayTickInt::name() const
		{
			Name result(this->_name.c_str());
			return result;
		}

		size_t ElementIntradayTickInt::numValues() const
		{
			return 1;
		}

		size_t ElementIntradayTickInt::numElements() const
		{
			return 0;
		}

		bool ElementIntradayTickInt::isArray() const
		{
			return false;
		}

		bool ElementIntradayTickInt::isComplexType() const
		{
			return false;
		}
		
		bool ElementIntradayTickInt::hasElement(const char* name, bool excludeNullElements) const
		{
			return false;
		}

		std::ostream& ElementIntradayTickInt::print(std::ostream& stream, int level, int spacesPerLevel) const
		{
			this->prettyPrintHelper(stream, level, spacesPerLevel, this->_value);
			return stream;
		}

		int ElementIntradayTickInt::getValueAsInt32(int index) const
		{
			if(index == 0)
				return this->_value;
			else
				throw elementPtrEx;
		}

		long ElementIntradayTickInt::getValueAsInt64(int index) const
		{
			if(index == 0)
				return this->_value;
			else
				throw elementPtrEx;
		}
		
		float ElementIntradayTickInt::getValueAsFloat32(int index) const
		{
			if(index == 0)
				return (float)this->_value;
			else
				throw elementPtrEx;
		}

		double ElementIntradayTickInt::getValueAsFloat64(int index) const
		{
			if(index == 0)
				return this->_value;
			else
				throw elementPtrEx;
		}

		const char * ElementIntradayTickInt::getValueAsString(int index) const
		{
			if(index == 0)
			{
				const char * result = ElementPtr::toCharPointer(this->_value);
				return result;
			}
			else
				throw elementPtrEx;
		}



	}
}